index/internal/indexer/indexer.go

```diff
@@ -0,0 +1,302 @@
+				adj[i] = append(adj[i], j)
+				adj[j] = append(adj[j], i)
+			}
+		}
+	}
+
+	// DFS to find connected components
+	visited := make([]bool, n)
+	for i := range n {
+		if visited[i] {
+			continue
+		}
+		stack := []int{i}
+		component := make(collections.Set[*ModuleTarget])
+		visited[i] = true
+
+		for len(stack) > 0 {
+			curr := stack[len(stack)-1]
+			stack = stack[:len(stack)-1]
+			component[targets[curr]] = true
+			for _, neighbor := range adj[curr] {
+				if !visited[neighbor] {
+					stack = append(stack, neighbor)
+					visited[neighbor] = true
+				}
+			}
+		}
+		groups = append(groups, component)
+	}
+	return groups
+}
+
+// Given set of targets that define the same headers try to select ones that contain other targets as their direct or transitive dependencies
+func SelectRootTargets(targets collections.Set[*ModuleTarget]) []*ModuleTarget {
+	allTargets := make(map[label.Label]*ModuleTarget)
+	dependentTargets := make(collections.Set[label.Label])
+
+	// Collect all target names
+	for target := range targets {
+		allTargets[target.Name] = target
+	}
+
+	// Mark all targets that are listed as dependencies
+	for target := range targets {
+		for dep := range target.Deps {
+			dependentTargets[dep] = true
+		}
+	}
+
+	// Any target not in the dependency map is a root
+	roots := make(collections.Set[*ModuleTarget])
+	for name, target := range allTargets {
+		if !dependentTargets[name] {
+			roots.Add(target)
+		}
+	}
+
+	return roots.Values()
+}
+
+func shouldExcludeHeader(path string) bool {
+	// Exclude blank paths.
+	if strings.TrimSpace(path) == "" {
+		return true
+	}
+
+	// Exlucde possily hidden files
+	segments := filepath.SplitList(path)
+	for _, segment := range segments {
+		if strings.HasPrefix(segment, ".") || strings.HasPrefix(segment, "_") {
+			return true
+		}
+	}
+	return false
+}
+
+// shouldExcludeTarget determines if the given target (label) is possibly internal.
+func shouldExcludeTarget(label label.Label) bool {
+	// Check target's path segments: if any segment (split on non-word characters and filtered to letters)
+	for _, segment := range filepath.SplitList(label.Pkg) {
+		tokens := splitWords(segment)
+		for _, token := range tokens {
+			switch token {
+			case "internal", "impl":
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// splits a string on non-letter characters.
+func splitWords(s string) []string {
+	isLetter := func(r rune) bool {
+		return (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z')
+	}
+	tokens := strings.FieldsFunc(s, func(r rune) bool {
+		return !isLetter(r)
+	})
+	// Filter out any empty tokens.
+	var result []string
+	for _, t := range tokens {
+		if t != "" {
+			result = append(result, t)
+		}
+	}
+	return result
+}
+
+// normalizeHeaderPath applies several normalization steps to make the header path
+// conform to a format that a C compiler or cc_rules can correctly resolve.
+func normalizeHeaderPath(hdrPath string, target ModuleTarget) string {
```

Comment:
If we're normalizing headers, I'd suggest using `filepath.ToSlash` to avoid working with backslash paths on Windows.